File: jawa/venator/src/main/java/ru/prohor/universe/venator/webhook/SignatureService.java
package ru.prohor.universe.venator.webhook;

import org.apache.commons.codec.digest.HmacAlgorithms;
import org.apache.commons.codec.digest.HmacUtils;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;

@Service
public class SignatureService {
    private final HmacUtils hmacUtils;

    public SignatureService(@Value("${universe.venator.webhook.secret}") String secret) {
        this.hmacUtils = new HmacUtils(HmacAlgorithms.HMAC_SHA_256, secret);
    }

    public boolean verifySignature(String signature, String body) {
        if (signature == null)
            return false;
        String hmacHex = hmacUtils.hmacHex(body);
        String expectedSignature = "sha256=" + hmacHex;
        return signature.equals(expectedSignature);
    }
}

File: jawa/korriban/venator-kt/src/main/kotlin/webhook/SignatureService.kt
package ru.prohor.universe.venator.webhook

import org.apache.commons.codec.digest.HmacAlgorithms
import org.apache.commons.codec.digest.HmacUtils
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Service

@Service
class SignatureService(
    @Value($$"${universe.venator.webhook.secret}") secret: String
) {
    private val hmacUtils: HmacUtils = HmacUtils(HmacAlgorithms.HMAC_SHA_256, secret)

    fun verifySignature(signature: String?, body: String): Boolean {
        return signature?.let { it == "sha256=${hmacUtils.hmacHex(body)}" } ?: false
    }
}
